File: packages/android-client/app/src/main/java/com/mycamerascontroller/client/holo/Motion.kt
package com.mycamerascontroller.client.holo

import android.view.Choreographer
import kotlin.math.abs
import kotlin.math.ceil
import kotlin.math.exp
import kotlin.math.min

/**
 * Procedural motion for Android views.
 *
 * The same rule as the desktop client: no interpolators, no fixed durations.
 * Every animated value is integrated from a physical model each frame, which
 * is what lets a press be interrupted by a fling, or a card be re-targeted
 * halfway through its own entrance, without a visible seam.
 */

class Spring(value: Float = 0f, var stiffness: Float = 170f, var damping: Float = 22f) {
    var value = value
    var velocity = 0f
    var target = value

    fun set(v: Float) { value = v; target = v; velocity = 0f }
    fun to(v: Float) { target = v }
    fun kick(v: Float) { velocity += v }

    fun update(dt: Float): Float {
        // Sub-stepped: a stiff spring and a dropped frame is otherwise all it
        // takes to launch a view off the screen.
        val steps = min(6, ceil(dt / (1f / 120f)).toInt().coerceAtLeast(1))
        val h = dt / steps
        repeat(steps) {
            val accel = (target - value) * stiffness - velocity * damping
            velocity += accel * h
            value += velocity * h
        }
        return value
    }

    val settled: Boolean get() = abs(value - target) < 0.0005f && abs(velocity) < 0.0005f
}

class Damper(value: Float = 0f, var rate: Float = 10f) {
    var value = value
    var target = value
    fun set(v: Float) { value = v; target = v }
    fun to(v: Float) { target = v }
    fun update(dt: Float): Float {
        value = target + (value - target) * exp(-rate * dt)
        return value
    }
}

/**
 * One Choreographer callback for the whole interface. A per-view animator
 * would mean dozens of independent callbacks fighting for the same frame;
 * this keeps ordering deterministic and makes adding an animation free.
 */
object HoloTicker : Choreographer.FrameCallback {
    private val callbacks = LinkedHashSet<(Float, Float) -> Unit>()
    private val pending = ArrayList<(Float, Float) -> Unit>()
    private val removing = ArrayList<(Float, Float) -> Unit>()
    private var lastNanos = 0L
    private var running = false
    var time = 0f; private set

    fun add(cb: (Float, Float) -> Unit): () -> Unit {
        pending.add(cb)
        start()
        return { removing.add(cb) }
    }

    private fun start() {
        if (running) return
        running = true
        lastNanos = 0L
        Choreographer.getInstance().postFrameCallback(this)
    }

    fun stop() {
        running = false
        Choreographer.getInstance().removeFrameCallback(this)
    }

    override fun doFrame(frameTimeNanos: Long) {
        if (!running) return

        if (pending.isNotEmpty()) { callbacks.addAll(pending); pending.clear() }
        if (removing.isNotEmpty()) { callbacks.removeAll(removing.toSet()); removing.clear() }

        // Nothing left watching the clock: stop rather than posting an empty
        // callback every 16ms for the rest of the process's life. The next
        // add() restarts the loop.
        if (callbacks.isEmpty()) { running = false; lastNanos = 0L; return }
        Choreographer.getInstance().postFrameCallback(this)

        val dt = if (lastNanos == 0L) 1f / 60f
        else ((frameTimeNanos - lastNanos) / 1e9f).coerceIn(1f / 240f, 1f / 20f)
        lastNanos = frameTimeNanos
        time += dt

        // Iterate a snapshot: a callback is allowed to add or remove others.
        for (cb in callbacks.toList()) cb(dt, time)
    }
}

/** Glyph scramble used for every label that changes at runtime. */
class Scramble(private val speed: Float = 1f) {
    private val glyphs = "АБВГДЕЖЗИКЛМНОПРСТУФХЦЧШЩЭЮЯ0123456789#%&@*<>/\\|=+-"
    private var frame = 0f
    private var stop: (() -> Unit)? = null
    private var current = ""
    private var queue: List<Item> = emptyList()

    private class Item(val from: Char?, val to: Char?, val start: Int, val end: Int) {
        var glyph: Char? = null
    }

    fun setInstant(text: String, apply: (String) -> Unit) {
        stop?.invoke(); stop = null
        current = text
        apply(text)
    }

    fun set(text: String, apply: (String) -> Unit) {
        if (text == current) return
        val from = current
        current = text
        val length = maxOf(from.length, text.length)
        queue = (0 until length).map { i ->
            val start = (Math.random() * 12 / speed).toInt()
            Item(from.getOrNull(i), text.getOrNull(i), start, start + (Math.random() * 14 / speed).toInt() + 4)
        }
        frame = 0f
        stop?.invoke()
        stop = HoloTicker.add { dt, _ -> tick(dt, apply) }
    }

    private fun tick(dt: Float, apply: (String) -> Unit) {
        val sb = StringBuilder()
        var complete = 0
        for (item in queue) {
            when {
                frame >= item.end -> { complete++; item.to?.let { sb.append(it) } }
                frame >= item.start -> {
                    if (item.glyph == null || Math.random() < 0.3) item.glyph = glyphs.random()
                    sb.append(item.glyph)
                }
                else -> item.from?.let { sb.append(it) }
            }
        }
        apply(sb.toString())
        // Progress in nominal 60 Hz frames derived from real time, so the
        // effect runs at one speed on every device.
        frame += dt * 60f
        if (complete == queue.size) {
            stop?.invoke(); stop = null
            apply(current)
        }
    }

    fun dispose() { stop?.invoke(); stop = null }
}

File: packages/client/src/renderer/holo/ui/motion.ts
import { clamp, damp } from '../math.js';

/**
 * Procedural motion for DOM.
 *
 * There is not a single CSS transition or keyframe timing function driving
 * interface state in this app. Everything that moves is integrated every frame
 * from a physical model, because that is the only way a value can be
 * retargeted mid-flight — hover away before the hover finished, close a
 * panel while it is still opening — without a visible discontinuity.
 */

export class Spring {
  value: number;
  velocity = 0;
  target: number;

  constructor(value = 0, public stiffness = 170, public damping = 22) {
    this.value = value;
    this.target = value;
  }

  set(value: number) {
    this.value = value;
    this.target = value;
    this.velocity = 0;
  }

  to(target: number) { this.target = target; }

  /** Adds an instantaneous impulse — the physical way to express a "kick". */
  kick(velocity: number) { this.velocity += velocity; }

  update(dt: number): number {
    // Sub-stepped: a stiff spring integrated with one big step at 20 fps
    // will overshoot to infinity, and dropped frames are not hypothetical.
    const steps = Math.min(6, Math.max(1, Math.ceil(dt / (1 / 120))));
    const h = dt / steps;
    for (let i = 0; i < steps; i++) {
      const accel = (this.target - this.value) * this.stiffness - this.velocity * this.damping;
      this.velocity += accel * h;
      this.value += this.velocity * h;
    }
    return this.value;
  }

  get settled() {
    return Math.abs(this.value - this.target) < 0.0005 && Math.abs(this.velocity) < 0.0005;
  }
}

/** Exponential approach — for values where overshoot would be wrong. */
export class Damper {
  value: number;
  target: number;
  constructor(value = 0, public rate = 10) {
    this.value = value;
    this.target = value;
  }
  to(target: number) { this.target = target; }
  set(value: number) { this.value = value; this.target = value; }
  update(dt: number): number {
    this.value = damp(this.value, this.target, this.rate, dt);
    return this.value;
  }
}

/**
 * A shared ticker. One requestAnimationFrame for the whole DOM layer keeps
 * the ordering deterministic and means adding an animation costs nothing.
 */
export class Ticker {
  private readonly callbacks = new Set<(dt: number, time: number) => void>();
  private last = 0;
  private handle = 0;
  private running = false;
  time = 0;

  add(cb: (dt: number, time: number) => void): () => void {
    this.callbacks.add(cb);
    return () => this.callbacks.delete(cb);
  }

  start() {
    if (this.running) return;
    this.running = true;
    this.last = performance.now();
    const loop = (now: number) => {
      if (!this.running) return;
      this.handle = requestAnimationFrame(loop);
      const dt = clamp((now - this.last) / 1000, 1 / 240, 1 / 20);
      this.last = now;
      this.time += dt;
      for (const cb of this.callbacks) cb(dt, this.time);
    };
    this.handle = requestAnimationFrame(loop);
  }

  stop() {
    this.running = false;
    cancelAnimationFrame(this.handle);
  }
}

export const ticker = new Ticker();

const GLYPHS = 'АБВГДЕЖЗИКЛМНОПРСТУФХЦЧШЩЭЮЯ0123456789#%&@*<>/\\|=+-';

/**
 * Decodes text one glyph at a time out of a scrambling cipher.
 *
 * Used for every label that changes at runtime. It is not decoration: a
 * value that scrambles and resolves is impossible to miss, so status
 * changes never need a flash or a colour alarm to be noticed.
 */
export class ScrambleText {
  /** Progress in nominal 60 Hz frames, accumulated from real time so the
      effect runs at the same speed whatever the render is managing. */
  private frame = 0;
  private queue: Array<{ from: string; to: string; start: number; end: number; glyph: string }> = [];
  private stop: (() => void) | null = null;
  private current = '';

  constructor(private readonly el: HTMLElement, private readonly speed = 1) {}

  setInstant(text: string) {
    this.stop?.();
    this.stop = null;
    this.queue = [];
    this.current = text;
    this.el.textContent = text;
  }

  set(text: string) {
    if (text === this.current) return;
    const from = this.current;
    this.current = text;
    const length = Math.max(from.length, text.length);
    this.queue = [];
    for (let i = 0; i < length; i++) {
      const start = Math.floor(Math.random() * 12 / this.speed);
      const end = start + Math.floor(Math.random() * 14 / this.speed) + 4;
      this.queue.push({ from: from[i] ?? '', to: text[i] ?? '', start, end, glyph: '' });
    }
    this.frame = 0;
    this.stop?.();
    this.stop = ticker.add((dt) => this.tick(dt));
  }

  private tick(dt: number) {
    let output = '';
    let complete = 0;
    for (const item of this.queue) {
      if (this.frame >= item.end) {
        complete++;
        output += item.to;
      } else if (this.frame >= item.start) {
        if (!item.glyph || Math.random() < 0.3) {
          item.glyph = GLYPHS[Math.floor(Math.random() * GLYPHS.length)];
        }
        output += item.glyph;
      } else {
        output += item.from;
      }
    }
    this.el.textContent = output;
    this.frame += dt * 60;
    if (complete === this.queue.length) {
      this.stop?.();
      this.stop = null;
      this.el.textContent = this.current;
    }
  }

  dispose() { this.stop?.(); }
}

/**
 * Splits an element's text into per-glyph spans so each can be animated
 * independently. Returns the spans in document order.
 */
export function splitGlyphs(el: HTMLElement, text: string): HTMLSpanElement[] {
  el.textContent = '';
  const spans: HTMLSpanElement[] = [];
  for (const ch of text) {
    const span = document.createElement('span');
    span.className = 'glyph';
    // A space with no glyph collapses; a non-breaking space keeps the
    // rhythm of the line intact while still animating.
    span.textContent = ch === ' ' ? ' ' : ch;
    span.style.setProperty('--glyph-index', String(spans.length));
    el.appendChild(span);
    spans.push(span);
  }
  return spans;
}
